File: hope_ws/src/calibration/src/calibration.cpp
#include "calibration.h"

#include <algorithm>
#include <cmath>
#include <fstream>
#include <sstream>
#include <stdexcept>
#include <string>

namespace calibration {

namespace {

double median(std::vector<double> v) {
  if (v.empty()) {
    return 0.0;
  }
  std::sort(v.begin(), v.end());
  size_t n = v.size();
  if (n % 2 == 0) {
    return 0.5 * (v[n / 2 - 1] + v[n / 2]);
  }
  return v[n / 2];
}

}  // namespace

common::BallPhysics calibrateBallPhysics(
  const std::vector<std::vector<double>> & timestamps,
  const std::vector<std::vector<Eigen::Vector3d>> & trajectories,
  const Eigen::Vector3d & g)
{
  std::vector<double> drag_a;
  std::vector<double> drag_v2;
  std::vector<double> h_rest;
  std::vector<double> v_rest;

  for (size_t idx = 0; idx < trajectories.size(); ++idx) {
    const auto & pos = trajectories[idx];
    const auto & t = timestamps[idx];
    if (t.size() < 4) {
      continue;
    }
    std::vector<double> dt_arr(t.size() - 1);
    std::vector<Eigen::Vector3d> vel(pos.size() - 1);
    for (size_t i = 0; i + 1 < t.size(); ++i) {
      dt_arr[i] = t[i + 1] - t[i];
      vel[i] = (pos[i + 1] - pos[i]) / dt_arr[i];
    }
    std::vector<double> t_vel(vel.size());
    for (size_t i = 0; i < vel.size(); ++i) {
      t_vel[i] = 0.5 * (t[i] + t[i + 1]);
    }
    std::vector<double> dt_vel(vel.size() - 1);
    for (size_t i = 0; i + 1 < vel.size(); ++i) {
      dt_vel[i] = t_vel[i + 1] - t_vel[i];
    }
    std::vector<Eigen::Vector3d> acc(vel.size() - 1);
    for (size_t i = 0; i + 1 < vel.size(); ++i) {
      acc[i] = (vel[i + 1] - vel[i]) / dt_vel[i];
    }
    std::vector<Eigen::Vector3d> vel_mid(vel.size() - 1);
    for (size_t i = 0; i + 1 < vel.size(); ++i) {
      vel_mid[i] = 0.5 * (vel[i] + vel[i + 1]);
    }
    std::vector<double> z_mid(acc.size());
    for (size_t i = 0; i < acc.size(); ++i) {
      z_mid[i] = 0.5 * (pos[i].z() + pos[i + 2].z());
    }

    for (size_t i = 0; i < acc.size(); ++i) {
      if (z_mid[i] > 0.03) {
        Eigen::Vector3d a_minus_g = acc[i] - g;
        drag_a.push_back(a_minus_g.norm());
        drag_v2.push_back(std::pow(vel_mid[i].norm(), 2));
      }
    }

    for (size_t i = 1; i + 1 < pos.size(); ++i) {
      double z_prev = pos[i - 1].z();
      double z_curr = pos[i].z();
      double z_next = pos[i + 1].z();
      if (z_prev > 0.01 && z_curr < 0.005 && z_next > 0.01) {
        size_t idx_pre = (i >= 2) ? (i - 2) : 0;
        size_t idx_post = std::min(i + 1, vel.size() - 1);
        if (idx_pre < vel.size() && idx_post < vel.size()) {
          Eigen::Vector3d v_pre = vel[idx_pre];
          Eigen::Vector3d v_post = vel[idx_post];
          double v_h_pre = std::hypot(v_pre.x(), v_pre.y());
          double v_h_post = std::hypot(v_post.x(), v_post.y());
          if (v_h_pre > 0.1) {
            h_rest.push_back(v_h_post / v_h_pre);
          }
          if (std::abs(v_pre.z()) > 0.1) {
            v_rest.push_back(std::abs(v_post.z()) / std::abs(v_pre.z()));
          }
        }
      }
    }
  }

  double k;
  if (!drag_a.empty()) {
    double sum_av = 0.0, sum_v2 = 0.0;
    for (size_t i = 0; i < drag_a.size(); ++i) {
      sum_av += drag_a[i] * drag_v2[i];
      sum_v2 += drag_v2[i] * drag_v2[i];
    }
    k = (sum_v2 > 0.0) ? (sum_av / sum_v2) : common::BallPhysics().k;
  } else {
    k = common::BallPhysics().k;
  }

  common::BallPhysics defaults;
  double C_h = h_rest.empty() ? defaults.C_h : median(h_rest);
  double C_v = v_rest.empty() ? defaults.C_v : median(v_rest);

  common::BallPhysics result;
  result.k = k;
  result.C_h = C_h;
  result.C_v = C_v;
  return result;
}

CsvTrajectory loadTrajectoryCsv(const std::string & path) {
  std::ifstream fh(path);
  if (!fh) {
    throw std::runtime_error("Cannot open trajectory CSV: " + path);
  }
  CsvTrajectory out;
  std::string line;
  while (std::getline(fh, line)) {
    if (line.empty()) {
      continue;
    }
    std::stringstream ss(line);
    std::string tok;
    std::vector<std::string> cols;
    while (std::getline(ss, tok, ',')) {
      cols.push_back(tok);
    }
    if (cols.size() < 4) {
      continue;
    }
    std::string first = cols[0];
    std::transform(first.begin(), first.end(), first.begin(),
      [](unsigned char c) { return std::tolower(c); });
    if (first == "t" || first == "time") {
      continue;
    }
    out.t.push_back(std::stod(cols[0]));
    out.p.emplace_back(std::stod(cols[1]), std::stod(cols[2]), std::stod(cols[3]));
  }
  return out;
}

}  // namespace calibration

File: hope_ws/src/hope_planner/hope_planner/calibration.py
"""Ball-physics calibration (HOPE planner step 8).

Fit the drag coefficient k and the horizontal/vertical restitution
coefficients C_h, C_v from recorded ball trajectories, following
HOPE_7DOF_Racket_Model_based_Planner_Reference_Setup.md, Section 2.1.

`calibrate_ball_physics()` is the pure-numpy fitter from the reference doc.
A small CLI (`hope_calibrate`) loads recorded trajectories from CSV files
(one file per trajectory: columns t,x,y,z) and prints the fitted parameters.
"""

import argparse
import csv
from typing import List

import numpy as np

from .constants import BallPhysics


def calibrate_ball_physics(
    trajectories: List[np.ndarray],
    timestamps: List[np.ndarray],
    g: np.ndarray = np.array([0.0, 0.0, -9.81]),
) -> BallPhysics:
    """Calibrate k, C_h, C_v from recorded ball trajectories.

    Parameters
    ----------
    trajectories : list of (N, 3) arrays
        Each array is a sequence of ball positions [x, y, z] in the HOPE frame.
    timestamps : list of (N,) arrays
        Corresponding timestamps in seconds.

    Returns
    -------
    BallPhysics with fitted parameters.
    """
    drag_a = []
    drag_v2 = []
    h_restitution = []
    v_restitution = []

    for pos, t in zip(trajectories, timestamps):
        if len(t) < 4:
            continue
        dt = np.diff(t)
        vel = np.diff(pos, axis=0) / dt[:, None]
        t_vel = 0.5 * (t[:-1] + t[1:])
        dt_vel = np.diff(t_vel)
        acc = np.diff(vel, axis=0) / dt_vel[:, None]
        vel_mid = 0.5 * (vel[:-1] + vel[1:])
        z_mid = 0.5 * (pos[:-2, 2] + pos[2:, 2])

        # Flight samples: z well above the table.
        flight_mask = z_mid > 0.03
        if np.any(flight_mask):
            a_flight = acc[flight_mask]
            v_flight = vel_mid[flight_mask]
            a_minus_g = a_flight - g[None, :]
            drag_a.extend(np.linalg.norm(a_minus_g, axis=1))
            drag_v2.extend(np.linalg.norm(v_flight, axis=1) ** 2)

        # Bounce detection: three-sample pattern (descend -> contact -> rise).
        z = pos[:, 2]
        for i in range(1, len(z) - 1):
            if z[i - 1] > 0.01 and z[i] < 0.005 and z[i + 1] > 0.01:
                idx_pre = max(0, i - 2)
                idx_post = min(len(vel) - 1, i + 1)
                if idx_pre < len(vel) and idx_post < len(vel):
                    v_pre = vel[idx_pre]
                    v_post = vel[idx_post]
                    v_h_pre = np.hypot(v_pre[0], v_pre[1])
                    v_h_post = np.hypot(v_post[0], v_post[1])
                    if v_h_pre > 0.1:
                        h_restitution.append(v_h_post / v_h_pre)
                    if abs(v_pre[2]) > 0.1:
                        v_restitution.append(abs(v_post[2]) / abs(v_pre[2]))

    # Fit drag: |a - g| = k * |v|^2  (least squares through the origin).
    if drag_a:
        a_mag = np.array(drag_a)
        v2 = np.array(drag_v2)
        k = float(np.dot(a_mag, v2) / np.dot(v2, v2))
    else:
        k = 0.1261   # venue-fit default (see constants.BallPhysics)

    C_h = float(np.median(h_restitution)) if h_restitution else 0.64
    C_v = float(np.median(v_restitution)) if v_restitution else 0.9215

    return BallPhysics(k=k, C_h=C_h, C_v=C_v)


def load_trajectory_csv(path: str):
    """Load a single trajectory CSV with columns t,x,y,z (header optional)."""
    t_list, p_list = [], []
    with open(path, newline="") as fh:
        reader = csv.reader(fh)
        for row in reader:
            if not row or row[0].strip().lower() in ("t", "time"):
                continue  # skip header / blank
            t_list.append(float(row[0]))
            p_list.append([float(row[1]), float(row[2]), float(row[3])])
    return np.array(t_list), np.array(p_list)


def main(argv=None):
    parser = argparse.ArgumentParser(
        description="Calibrate HOPE ball physics (k, C_h, C_v) from recorded CSV trajectories."
    )
    parser.add_argument(
        "csv", nargs="+",
        help="One or more trajectory CSV files (columns: t,x,y,z in the HOPE frame).",
    )
    args = parser.parse_args(argv)

    trajectories, timestamps = [], []
    for path in args.csv:
        t, p = load_trajectory_csv(path)
        timestamps.append(t)
        trajectories.append(p)

    phys = calibrate_ball_physics(trajectories, timestamps)
    print("Calibrated ball physics:")
    print(f"  drag_k        = {phys.k:.4f}")
    print(f"  restitution_h = {phys.C_h:.4f}")
    print(f"  restitution_v = {phys.C_v:.4f}")
    print("\nPaste into config/hope_planner.yaml:")
    print(f"    drag_k: {phys.k:.4f}")
    print(f"    restitution_h: {phys.C_h:.4f}")
    print(f"    restitution_v: {phys.C_v:.4f}")


if __name__ == "__main__":
    main()
